src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -356,52 +356,56 @@ protected void CheckDocumentIsOpen(DocumentId documentId)
         }
 
         protected internal void OnDocumentOpened(
-            DocumentId documentId, SourceTextContainer textContainer,
-            bool isCurrentContext = true)
+            DocumentId documentId, SourceTextContainer textContainer, bool isCurrentContext = true)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                CheckDocumentIsInCurrentSolution(documentId);
-                CheckDocumentIsClosed(documentId);
+            SetCurrentSolution(
+                static (oldSolution, data) =>
+                {
+                    var (@this, documentId, _, _) = data;
 
-                var oldSolution = this.CurrentSolution;
-                var oldDocument = oldSolution.GetRequiredDocument(documentId);
-                var oldDocumentState = oldDocument.State;
+                    CheckDocumentIsInSolution(oldSolution, documentId);
+                    @this.CheckDocumentIsClosed(documentId);
 
-                AddToOpenDocumentMap(documentId);
```

Comment:
this code was slightly weird.  it added thsi prior to actually updating the workspace. that didn't seem necessary, so i just moved to this to the first statement of onAfterUpdate.